substrate/primitives/virtualization/src/lib.rs

```diff
@@ -0,0 +1,226 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! This crate is intended for use by runtime code (e.g pallet-contracts) to spawn PolkaVM instances
```

Comment:
No. This will be specifically targeted at smart contract usecases and PolkaVM. We have no plans for spawning WASM VMs; we already tried that in the past, and it was a bad idea/it didn't work.

Reply:
Well, historically we had three features which are now removed:

1) `wasmi` as an executor; this was meant to be used to run *full* runtimes as a sort of a backup executor to `wasmtime`, but since `wasmtime` got rock solid we removed it,
2) we wanted to use the executor infrastructure for running smart contracts on `wasmtime`/`wasmer`/whatever in an accelerated manner, but it ended up being a bad idea (because the VMs weren't really designed for this use case) so we removed it,
3) we had host calls to spawn new executors from within the runtime and run something in them, but the implementation for this was horribly broken *and* no one used it, so we removed it.

So right now essentially we're only looking into (2), but this time with a VM that is designed to support this use case.

(The (3) might actually be possible with Polkajam, but it's still too early to talk about that as we're still designing/prototyping it.)

Reply:
There is also a 2a) where we actually had a predecessor of this API where we spawned wasmi instances for smart contracts. But it turned out to not be the hassle and we moved wasmi into the runtime. 

However, PolkaVM can't live inside the runtime as a JIT.

Wasm smart contracts will be run with an in-runtime wasmi. There is no plan to support anything else than PolkaVM with this virtualization API.

Reply:
> However, PolkaVM can't live inside because it compiles to native instructions.

If runtime itself is RISC-V it should be able to, also interpreted mode is still available as a fallback in polkavm, isn't it?